client-runtime/src/main/java/com/microsoft/rest/v2/protocol/HttpResponseDecoder.java

```diff
@@ -0,0 +1,291 @@
+ * license information.
+ */
+
+package com.microsoft.rest.v2.protocol;
+
+import com.fasterxml.jackson.core.JsonParseException;
+import com.google.common.reflect.TypeToken;
+import com.microsoft.rest.v2.Base64Url;
+import com.microsoft.rest.v2.DateTimeRfc1123;
+import com.microsoft.rest.v2.RestResponse;
+import com.microsoft.rest.v2.SwaggerMethodParser;
+import com.microsoft.rest.v2.UnixTime;
+import com.microsoft.rest.v2.http.HttpHeaders;
+import com.microsoft.rest.v2.http.HttpResponse;
+import com.microsoft.rest.v2.util.FlowableUtil;
+import io.reactivex.Completable;
+import io.reactivex.Maybe;
+import io.reactivex.Observable;
+import io.reactivex.Single;
+import io.reactivex.functions.Function;
+import org.joda.time.DateTime;
+
+import java.io.IOException;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Deserializes an {@link HttpResponse}.
+ */
+public final class HttpResponseDecoder {
+    private final SwaggerMethodParser methodParser;
+    private final SerializerAdapter<?> serializer;
+
+    /**
+     * Creates an HttpResponseDecoder.
+     * @param methodParser metadata about the Swagger method used for decoding.
+     * @param serializer the serializer
+     */
+    public HttpResponseDecoder(SwaggerMethodParser methodParser, SerializerAdapter<?> serializer) {
+        this.methodParser = methodParser;
+        this.serializer = serializer;
+    }
+
+    /**
+     * Asynchronously decodes an {@link HttpResponse}, deserializing into a response or error value.
+     * @param response the {@link HttpResponse}
+     * @return A {@link Single} containing either the decoded HttpResponse or an error
+     */
+    public Single<HttpResponse> decode(final HttpResponse response) {
+        final Object deserializedHeaders;
+        try {
+            deserializedHeaders = deserializeHeaders(response.headers());
+        } catch (IOException e) {
+            return Single.error(e);
+        }
+
+        final Type returnValueWireType = methodParser.returnValueWireType();
+
+        final TypeToken entityTypeToken = getEntityType();
+
+        boolean isSerializableBody = !FlowableUtil.isFlowableByteArray(entityTypeToken)
+            && !entityTypeToken.isSubtypeOf(Completable.class)
+            && !entityTypeToken.isSubtypeOf(byte[].class)
+            && !entityTypeToken.isSubtypeOf(Boolean.TYPE) && !entityTypeToken.isSubtypeOf(Boolean.class)
+            && !entityTypeToken.isSubtypeOf(Void.TYPE) && !entityTypeToken.isSubtypeOf(Void.class);
+
+        int[] expectedStatuses = methodParser.expectedStatusCodes();
+        boolean isErrorStatus = true;
+        if (expectedStatuses != null) {
+            for (int i = 0; i < expectedStatuses.length && isErrorStatus; i++) {
+                if (expectedStatuses[i] == response.statusCode()) {
+                    isErrorStatus = false;
```

Comment:
Don't forget to `break` after the condition you want is encountered. It makes things a little quicker.

Reply:
This makes me think that simply putting a `break;` is going to be clearer to maintainers than `&& isErrorStatus` in the loop condition.